include/drivers/arm/ccn.h

```diff
@@ -76,6 +76,16 @@ typedef struct ccn_desc {
 	uintptr_t periphbase;
 } ccn_desc_t;
 
+/* Enum used to loop through all types of nodes in CCN*/
+typedef enum node_types {
+	NODE_TYPE_RNF = 0,
+	NODE_TYPE_RNI,
+	NODE_TYPE_RND,
+	NODE_TYPE_HNF,
+	NODE_TYPE_HNI,
+	NODE_TYPE_SN,
+	NUM_NODE_TYPES
+} node_types_t;
```

Comment:
In order to comply with MISRA it looks like you will have to turn this enum into a set of #defines. You are using this inside ``static unsigned int get_region_id_for_node(node_types_t node_type, unsigned int node_id)``: ``(node_type << 4)``

I'm afraid that MISRA doesn't like enums being used in arithmetic/logical operations.

Reply:
In the same file ccn.c, similar usage is present:

static unsigned int ccn_get_rn_master_info(uintptr_t periphbase,
                                           rn_info_t *info)
{
        unsigned int num_masters = 0;
        rn_types_t rn_type;

        assert (info);

        for (rn_type = RN_TYPE_RNF; rn_type < NUM_RN_TYPES; rn_type++) {
                unsigned int mn_reg_off, node_id;
                unsigned long long rn_bitmap;

                /*
                 * RN-F, RN-I, RN-D node registers in the MN region occupy
                 * contiguous 16 byte apart offsets.
                 */
                mn_reg_off = MN_RNF_NODEID_OFFSET + (rn_type << 4);
-------------------------------------------------------------------------------------------------

I guess, I am using the same way.

Reply:
Hmmm... Ok, I guess we need to do a cleanup of this driver anyway.